### PR TITLE
Don't store project files on docker image

### PR DESCRIPTION
**Pull request by tombruijn**
@stuartellis Great work on #857 ! Sorry I wasn't more active to review it.
I had a look and wondered why the image includes the bundle as that could introduce a stale state between the bundle on the image and the bundle of the project. Especially since it doesn't automatically build the image before running a command on a docker container.
To fix that I applied some of the setup I've recently setup for some other Docker testing projects.
Let me know if this sounds like a good idea to apply here too

Also wasn't sure why the `-I` and `-S` options were used for the Ruby scripts on the docker containers. Seems to work without it? If I missed some application of this, it might be nicer to move this setup to some of the files, so they're not reliant of the command that runs the test suites.

---

*commit message*:

This change only prepares the build environment on the docker image, but does
not create a production image which could be deployed for an app.

It uses volumes and a custom `BUNDLE_PATH` to store the gems in the
bundle which can be used between updates to the bundle itself. If a gem
version changes it would not have to reinstall all the gems and build
the image from that point, it would be able to reuse the bundler cache
(in the mounted volume) and only require an update to the changed gems.
Overal, should speed up the build process for the docker test
environment.

Added a `bin/docker_test` file that performs tasks within the docker
container.

Also made building the docker image mandatory when running the `docker`
namespaced rake files, so you can't end up with an out-of-sync state
where the docker image has changed, but hasn't been rebuild so a
developer would be testing against a stale image.

### CODE REVIEW

**Comment by stuartellis** [comment]
Thank you. Until yesterday, the Rake tasks did force a rebuild of the Docker image, but I second-guessed myself, and took this out!

TLDR;

* I am not a Docker expert, at all, so am inclined to defer to your experience

Agree:

* Should force rebuild of images
* Should get rid of the *ruby -I -S* thing

Needs more discussion:

* Caching of bundle on host seems slow
* Use of shell script to run tasks inside container (necessary if we need to run Bundler before each task, otherwise not keen)

Longer comments:

IIRC, the  -I -S were needed to make *something* work properly at some point, but it's been so long since I wrote the PR that I can't remember.

I like caching the bundle on the host, but it seems to make things much slower:

    $ rake docker:spec
    Finished in 16.17 seconds (files took 54.74 seconds to load)
    1411 examples, 0 failures

Compared to:
    
    $ rake docker:spec
    Finished in 5.64 seconds (files took 2.68 seconds to load)
    1411 examples, 0 failures

Performance is more of an issue for integration tests: IIRC, once I had ported several sets of these to Docker, the integration test task took over a minute to run on my development laptop.

Just on principle, I don't like a separate shell script running the tasks inside the container, since we actually have Rake available on both the host, and inside the container as well. Obviously, if we are going to trigger Bundler inside the container before each task, then it makes sense.

**Comment by tombruijn** [comment]
Well, we don't need to force run bundler beforehand. We can also put the "installation" step in a build task on the docker container. (It's what I actually do in most other projects.) Then it has the same behavior as most development environments; missing a gem?: bundler will error.

About the bash script, it's a personal preference of mine as it's quicker to run and doesn't require the bundle and such to be installed locally. You basically won't need any local setup to run the tasks. We can move the stuff from the bash script to rake tasks, but it would add additional load time locally as well.

I've improved the performance a bit by specifying volume modes, but it will always be a bit slower on mounted volumes. I haven't seen such a dramatic increase in load/boot/require time for any project, but then again, Backup is special with its many dependencies. Requiring all those files will take a while unfortunately, it's still Ruby.

Previous setup without volume modes configured (perfect consistency between host and container):

```
Finished in 16.99 seconds (files took 1 minute 1.66 seconds to load)
1411 examples, 0 failures

bundle exec rake docker:spec  1.17s user 0.33s system 1% cpu 1:23.29 total
```

New setup with volume modes configured (delayed file sync to host from container):

```
Finished in 8.75 seconds (files took 14.43 seconds to load)
1411 examples, 0 failures

bundle exec rake docker:spec  1.12s user 0.30s system 4% cpu 28.989 total
```

would that be an acceptable time?

**Comment by stuartellis** [comment]
Unfortunately, the latest version seems to be broken. I deleted images and containers, then ran commit 59fb427:

> $ rake docker:spec
> docker-compose build
> Building ruby_backup_tester
> Step 1/6 : FROM ruby:2.4.2
>  ---> eebb1381c2aa
> Step 2/6 : LABEL maintainer "stuart@stuartellis.name" version "0.1.0" description "Image for running the backup Rubygem"
>  ---> Running in 131ead29f86e
>  ---> b2067a52a34f
> Removing intermediate container 131ead29f86e
> Step 3/6 : ENV APP_DEPS bsdtar ca-certificates curl g++ git     libxml2 libxslt1.1 libyaml-0-2 openssl
>  ---> Running in 88e52817d672
>  ---> cfca0b8e0dbd
> Removing intermediate container 88e52817d672
> Step 4/6 : RUN apt-get update && apt-get install -y --no-install-recommends $APP_DEPS
>  ---> Running in 203da13cbe22
> Get:1 http://security.debian.org jessie/updates InRelease [63.1 kB]
> Ign http://deb.debian.org jessie InRelease
> Get:2 http://deb.debian.org jessie-updates InRelease [145 kB]
> Get:3 http://security.debian.org jessie/updates/main amd64 Packages [547 kB]
> Get:4 http://deb.debian.org jessie Release.gpg [2373 B]
> Get:5 http://deb.debian.org jessie Release [148 kB]
> Get:6 http://deb.debian.org jessie-updates/main amd64 Packages [23.1 kB]
> Get:7 http://deb.debian.org jessie/main amd64 Packages [9063 kB]
> Fetched 9992 kB in 17s (565 kB/s)
> Reading package lists...
> Reading package lists...
> Building dependency tree...
> Reading state information...
> ca-certificates is already the newest version.
> g++ is already the newest version.
> libxslt1.1 is already the newest version.
> libxslt1.1 set to manually installed.
> libyaml-0-2 is already the newest version.
> libyaml-0-2 set to manually installed.
> openssl is already the newest version.
> openssl set to manually installed.
> curl is already the newest version.
> git is already the newest version.
> libxml2 is already the newest version.
> libxml2 set to manually installed.
> The following extra packages will be installed:
>   libarchive13
> Suggested packages:
>   bsdcpio lrzip
> The following NEW packages will be installed:
>   bsdtar libarchive13
> 0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
> Need to get 325 kB of archives.
> After this operation, 795 kB of additional disk space will be used.
> Get:1 http://deb.debian.org/debian/ jessie/main libarchive13 amd64 3.1.2-11+deb8u3 [271 kB]
> Get:2 http://deb.debian.org/debian/ jessie/main bsdtar amd64 3.1.2-11+deb8u3 [53.9 kB]
> debconf: delaying package configuration, since apt-utils is not installed
> Fetched 325 kB in 1s (323 kB/s)
> Selecting previously unselected package libarchive13:amd64.
> (Reading database ... 21171 files and directories currently installed.)
> Preparing to unpack .../libarchive13_3.1.2-11+deb8u3_amd64.deb ...
> Unpacking libarchive13:amd64 (3.1.2-11+deb8u3) ...
> Selecting previously unselected package bsdtar.
> Preparing to unpack .../bsdtar_3.1.2-11+deb8u3_amd64.deb ...
> Unpacking bsdtar (3.1.2-11+deb8u3) ...
> Setting up libarchive13:amd64 (3.1.2-11+deb8u3) ...
> Setting up bsdtar (3.1.2-11+deb8u3) ...
> Processing triggers for libc-bin (2.19-18+deb8u10) ...
>  ---> 8f614a863bdc
> Removing intermediate container 203da13cbe22
> Step 5/6 : ENV APP_HOME /usr/src/backup
>  ---> Running in 10a9c90b1543
>  ---> 7fc196bfb296
> Removing intermediate container 10a9c90b1543
> Step 6/6 : WORKDIR $APP_HOME
>  ---> 94786417ef99
> Removing intermediate container a072f97ea7d1
> Successfully built 94786417ef99
> Successfully tagged backup/test-suite:local
> docker-compose run --rm ruby_backup_tester bin/docker_test rspec
> bundler: command not found: rspec
> Install missing gem executables with `bundle install`
> rake aborted!
> Command failed with status (127): [docker-compose run --rm ruby_backup_tester...]

**Comment by tombruijn** [comment]
@stuartellis did you run `rake docker:prepare` beforehand?
It's no longer automatically checking and installing missing gems. I can add that back if it's more convenient?

**Comment by stuartellis** [comment]
Aha - yes, if I use the tasks correctly, they work :)

    $ rake docker:prepare
    $ rake docker:spec

Says:

> Finished in 11.43 seconds (files took 28.48 seconds to load)
> 1411 examples, 0 failures

I would be inclined to just remove the _desc_ line for the _docker:build_ task, so that it doesn't appear on the list of tasks: the other tasks run it as a dependency, so it could effective be "private".

**Comment by stuartellis** [comment]
A small thing - the rename of the Docker image has broken the clobber task. It looks like the Docker commands aren't consistent:

> 
> $ rake docker:clobber
> $ docker images
> REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
> backup/test-suite   local               5dc416e3acd0        11 minutes ago      697MB
> 

Example output of ` docker ps`:

> 
> $ docker ps -a
> CONTAINER ID        IMAGE                     COMMAND                  CREATED             STATUS              PORTS               NAMES
> 7898458d6824        backup/test-suite:local   "bin/docker_test c..."   17 seconds ago      Up 15 seconds                           backup_ruby_backup_tester_run_1

**Comment by tombruijn** [comment]
> the rename of the Docker image has broken the clobber task. It looks like the Docker commands aren't consistent with each other

I changed the `docker run` command to include the `--rm` option, which means it will automatically remove the container after it finishes. So I'm not sure if the clean task is still necessary? Any reason to keep the containers after running the test suites?

**Comment by stuartellis** [comment]
Yes, I noticed that the containers are now automatically removed (which is great), and would be happy to remove the "clean" task. The "clobber" task actually deletes the images, and that doesn't work ATM: _docker images_ has a different format than the _grep_ assumes.
